### PR TITLE
 Acceptance tests: prevent deadlocks in tests for async writes

### LINKED ISSUE #26: Acceptance tests: prevent deadlocks in tests for async writes
### Feature description

In two of our tests for async writes, we are waiting for all the ack functions to be called. To prevent deadlocks, we need to specify a timeout, so that tests do not wait indefinitely, in cases where there's actually a bug in the connector code.

### PR DESCRIPTION
**Pull request by hariso**
### Description

Fixes #26 

### Quick checks:

- [x] I have followed the [Code Guidelines](https://github.com/ConduitIO/conduit/blob/main/docs/code_guidelines.md).
- [x] There is no other [pull request](https://github.com/ConduitIO/conduit-connector-sdk/pulls) for the same
  update/change.
- [ ] I have written unit tests.
- [x] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

File: acceptance_testing.go [hunk 1]
```diff
@@ -1129,3 +1129,17 @@ func (a acceptanceTest) isEqualData(is *is.I, want, got Data) {
 		is.Equal(want.Bytes(), got.Bytes()) // data did not match (want != got)
 	}
 }
+
+func waitTimeout(wg *sync.WaitGroup, timeout time.Duration) bool {
```
[the remark this answers is not in the document]
Comment (hariso) [reply acceptance_testing.go:1133]: I was thinking about that too, but I didn't like the duplication at all. On the other hand, maybe it's not a bad thing at all that it's accessible to the rest of the package. I'd maybe move it to a utils package or import it from a library.

**Review by dylanlott — COMMENTED** [review]
I saw the comment about pulling the wait utils functions into a separate file or package. I would be in favor of a separate package entirely over a file, especially if these functions are useful concurrency patterns for acceptance tests et. al. Looks good!

**Comment by lovromazgon** [comment]
Something was bugging me about these functions, bugging me to the point that I took the time to play around with a completely different approach. Here's what I came up with: https://github.com/ConduitIO/conduit-connector-sdk/commit/a9290b473be448836a4a0225afffa1215eda8307.

The idea is to have types that extend the functionality of `sync.WaitGroup` and `chan` with the functionality we need. Even if you use the original type you can cast it to our internal type and use the extended functionality, e.g.:
```
var wg sync.WaitGroup
err := (*internal.WaitGroup)(&wg).Wait(ctx)
// or
err := (*internal.WaitGroup)(&wg).WaitTimeout(ctx, time.Second)

c := make(chan struct{})
val, ok, err := internal.Chan[struct {}](c).Recv(ctx)
// or
val, ok, err := internal.Chan[struct {}](c).RecvTimeout(ctx)
```

Or you can use the internal types directly (as I did in the commit). WDYT @hariso?

**Comment by hariso** [comment]
@lovromazgon Nice! I've seen a similar approach in a library I was considering (introducing a dependency for a thing as simple as this wasn't really appealing). Your approach with types is more elegant though! Since we're basically writing our own WaitGroup alternative, we'd need tests for the new as well as the existing WaitGroup functionality. If we can agree the the scope change is OK and pays out, I think it's be cool seeing this merged. :+1:

**Comment by lovromazgon** [comment]
Glad to hear you like the approach.

> we'd need tests for the new as well as the existing WaitGroup functionality

I totally agree, I didn't include them since I made the commit more as an exploration to see if it made sense. Do you want me to write the tests or do you want to include the commit in your PR and write tests for it?

**Comment by hariso** [comment]
We're punting on #26 for now, so I'll close this.

**Pull request closed without merging**